cli/tools/repl.rs

```diff
@@ -211,71 +214,98 @@ impl Highlighter for Helper {
-          colors::gray(cap.as_str()).to_string()
-        } else if let Some(cap) = caps.name("keyword") {
-          colors::cyan(cap.as_str()).to_string()
-        } else if let Some(cap) = caps.name("infinity") {
-          colors::yellow(cap.as_str()).to_string()
-        } else if let Some(cap) = caps.name("classes") {
-          colors::green_bold(cap.as_str()).to_string()
-        } else if let Some(cap) = caps.name("hexnumber") {
-          colors::yellow(cap.as_str()).to_string()
-        } else if let Some(cap) = caps.name("octalnumber") {
-          colors::yellow(cap.as_str()).to_string()
-        } else if let Some(cap) = caps.name("binarynumber") {
-          colors::yellow(cap.as_str()).to_string()
-        } else {
-          caps[0].to_string()
-        }
-      })
-      .to_string()
-      .into()
+    let comments = SingleThreadedComments::default();
+    let lexer = Lexer::new(
+      ast::get_syntax(&MediaType::JavaScript),
+      ast::TARGET,
+      StringInput::new(line, BytePos(0), BytePos(line.len() as u32)),
+      Some(&comments),
+    );
+
+    let mut out_line = String::from(line);
+
+    // Adding color adds more bytes to the string,
+    // so an offset is needed to stop spans falling out of sync.
+    let mut offset = 0;
+
+    enum ItemKind {
+      Token(Token),
+      Comment(CommentKind),
+    }
+
+    struct Item {
+      span: Range<usize>,
+      kind: ItemKind,
+    }
+
+    fn span_to_range(span: Span) -> Range<usize> {
+      span.lo.0 as usize..span.hi.0 as usize
+    }
+
+    let mut items: Vec<Item> = Vec::new();
+    items.extend(lexer.map(|token| Item {
+      span: span_to_range(token.span),
+      kind: ItemKind::Token(token.token),
+    }));
+    items.extend(flatten_comments(comments).map(|comment| Item {
+      span: span_to_range(comment.span),
+      kind: ItemKind::Comment(comment.kind),
+    }));
+    items.sort_by_key(|item| item.span.start);
+
+    for item in items {
+      let before_len = out_line.len();
+      out_line.replace_range(
+        item.span.start + offset..item.span.end + offset,
+        &match item.kind {
+          ItemKind::Token(token) => match token {
+            Token::Str { .. } | Token::Template { .. } | Token::BackQuote => {
+              colors::green(&line[item.span]).to_string()
+            }
+            Token::Regex(_, _) => colors::red(&line[item.span]).to_string(),
+            Token::Num(_) => colors::yellow(&line[item.span]).to_string(),
+            Token::Word(word) => match word {
+              Word::True | Word::False | Word::Null => {
+                colors::yellow(&line[item.span]).to_string()
+              }
+              Word::Keyword(_) => colors::cyan(&line[item.span]).to_string(),
+              Word::Ident(js_word!("undefined")) => {
+                colors::gray(&line[item.span]).to_string()
+              }
+              Word::Ident(js_word!("Infinity"))
+              | Word::Ident(js_word!("NaN")) => {
```

Comment:
I think we could get away without using `js_word!` an instead calling `into()` on the strings

Reply:
Ok, fixed - `into()` didn't work, but `Atom` implements `PartialEq` for `str`.